### PR TITLE
Retry AWS S3 exceptions during initialisation

### PR DESCRIPTION
**Pull request by stojg**
This allows us to match the exceptions better before deciding to retry or not.

### CODE REVIEW

**Comment by stojg** [comment]
Not sure if I'm 100% liking this approach.. its a bit unreadable at some points.

**Comment by stojg** [comment]
I might also drag the retry out of Frisp into it's own package, so we can reuse it in our other tools..

File: Frisp/Initialiser.php [hunk 1]
```diff
@@ -98,9 +99,13 @@ protected function initDeputyRole($config) {
 			echo "\n\n";
 		}
 
-		$this->tryIgnoringException(
-			'Aws\Iam\Exception\MalformedPolicyDocumentException',
-			function() use($deputy, $iamClient, $realTrustPolicy) {
+		$retry = new Retry();
+		$retry->setIgnoredException('Aws\Iam\Exception\MalformedPolicyDocumentException');
+		$retry->setTimeout(30);
```
Comment (hafriedlander) [inline Frisp/Initialiser.php]: Why not make these (timeout & backoff) the default, and save yourself some typing?

File: Frisp/Initialiser.php [hunk 2]
```diff
@@ -98,9 +99,13 @@ protected function initDeputyRole($config) {
...
+		$retry->useExponentialBackoff();
+
+		while(!$retry->wasSuccess()) {
```
Comment (hafriedlander) [inline Frisp/Initialiser.php]: could do/while instead? Why is better than merging this and run into 'runUntilSuccess`?

File: Frisp/Initialiser.php [hunk 3]
```diff
@@ -257,17 +235,28 @@ protected function initBucket($config) {
...
 		// Aws\S3\Exception\S3Exception - "Invalid principal in policy" persists
-		$bucket = $this->env->bucket;
-		$encodedPolicy = json_encode($policy, JSON_PRETTY_PRINT);
-		$this->tryIgnoringException('Aws\S3\Exception\S3Exception', function() use($s3Client, $bucket, $encodedPolicy) {
-			$s3Client->putBucketPolicy([
-				'Bucket' => $bucket,
-				'Policy' => $encodedPolicy
-			]);
+		$retry = new Retry();
+		$retry->setExceptionMatcher(function(\Exception $e) {
```
Comment (hafriedlander) [inline Frisp/Initialiser.php]: exceptionMatcher seems to me to indicate a function that should return true or false.

File: Frisp/Initialiser.php (hunk 1 above)
Comment (stojg) [inline Frisp/Initialiser.php]: true, the timeout is actually defaulted to 32 secs, within (1 + 2 + 4 + 8 +16)

File: Frisp/Initialiser.php (hunk 3 above)
Comment (stojg) [inline Frisp/Initialiser.php]: what do you mean? It's a setter, what would that indicate? maybe change of name to .. um.. `setExceptionCatcher` or something similar?

File: Frisp/Initialiser.php (hunk 2 above)
Comment (stojg) [inline Frisp/Initialiser.php]: yeah, I was a bit unsure of what would be more readable, i made the assumption that `while .. do` is more familiar to most developer, where as `do ... while` is not used as often. But I not fussed anyway. The current one look a bit strange though.. "how can it be success "

File: Frisp/Initialiser.php (hunk 3 above)
Comment (hafriedlander) [inline Frisp/Initialiser.php]: `setExceptionMatcher` sets the "exception matcher", which to me would be a function that returns true if the exception "matches". Maybe `setOnExceptionCallback`?

Comment (stojg) [inline Frisp/Initialiser.php]: Hmm.. yeah.. let me think about this for a while..

**Pull request closed without merging**